components/sup/src/manager/service_updater/rolling_update_worker.rs

```diff
@@ -207,15 +256,20 @@ impl RollingUpdateWorker {
                                    self.service_group);
                             break FollowerUpdateStartEvent::PromotedToLeader;
                         }
-                        if leader.pkg != me.pkg {
+
+                        rolling_leader_state.update(leader);
+                        if rolling_leader_state.ok_to_lead() && leader.pkg != me.pkg {
```

Comment:
@mwrock , here's a scenario you may want to consider:

- Given 3 nodes A, B, C. A is the leader. B & C are followers.
- A, B & C currently have package version v1
- A updates the package to version v2
- The following happens concurrently:
  - A dies
  - package version v2 gets demoted 
- A new package version v3 is promoted  
- B becomes the new leader and updates to package version v3.

Now in C:
- The older leader is A, which had updated to package version v2
- The new leader is B, which has a package v3 which is not equal to v2

Thus C will always think that B is not ok to lead, since it's package version v3 does not match the version (v2) of the old leader (A). Thus C might always be stuck in this loop


## Using Lamport Clocks to track version change ordering

A possible solution to this problem, and also perhaps a general solution to the problem of rolling back updates across followers could be achieved using [Lamport clocks](https://martinfowler.com/articles/patterns-of-distributed-systems/lamport-clock.html) as follows:

- Given 3 nodes A, B, C. A is the leader. B & C are followers.
- A, B & C currently have package version 0/v1 . Here `0` serves as a lamport clock to track the order of version changes.
- A updated the package to 1/v2
- The following happens concurrently:
  - A dies
  - Package version v2 gets demoted
- A new package version v3 is promoted
- B becomes the new leader and updates the package version to 2/v3. Note that B has incremented the lamport clock from 1 to 2. 

Now in C:
- The older leader is A, which had updated to package version v2
- The new leader is B, which has package 2/v3 which is after 1/v2. Hence C will now update to the latest version which happens to be 2/v3.

The `ok_to_lead` function could be changed to check that the new leader's package version's lamport clock is newer than the old leader's version's lamport clock.

## Scenario of Package version Rollback

- Given 3 nodes A, B, C. A is the leader. B & C are followers.
- A, B & C currently have package version 0/v1.
- A updates to 1/v2.
- B updates to 1/v2 because it's previous peer (A) has updated to the leader's version (1/v2).
- Package version v2 gets demoted.
- A detects that  v1 is now the latest version and updates it's version to 2/v1.
- B & C receive the latest version (2/v1) of the package from the leader.
- C does not update, because it's previous peer (B) has version 1/v2, which is not equal to the leader's version 2/v1.
- B updates to 2/v1 because it's previous peer (A) has updated to the leader's version (2/v1). This is effectively a rollback. 
- C updates to 2/v1 because it's previous peer (B) has updated to the leader's (A) version (2/v1). C effectively does nothing in this case, as it transitions from 0/v1 -> 2/v2.

The use of a Lamport clock to track package update version changes removes all assumptions about how the package's version should progress. It could go forward or backward by any number of steps and still behave correctly.